File: src-tauri/src/media/video.rs
use std::{
    cell::LazyCell,
    collections::HashMap,
    ffi::OsStr,
    path::{Path, PathBuf},
    sync::{
        Arc,
        atomic::{AtomicBool, Ordering},
    },
};

use futures::stream::{self, StreamExt as FuturesStreamExt, TryStreamExt};
use serde::Deserialize;
use smol::{
    fs::{self, remove_file},
    io::{self},
    process::Command,
};
use which::which;

/// Video stream information
#[derive(Debug, Deserialize)]
struct Stream {
    codec_type: String,
    width: Option<i32>,
    height: Option<i32>,
    bit_rate: Option<String>,
}

/// Media file probe result
#[derive(Debug, Deserialize)]
struct MediaProbe {
    streams: Vec<Stream>,
}

/// Video information
#[derive(Debug, Clone)]
#[allow(unused)]
pub struct VideoInfo {
    width: i32,
    height: i32,
    bit_rate: i32,
}

/// Video processing preset configuration
#[derive(Debug, Clone)]
pub struct VideoPreset {
    /// Executor name (e.g., "ffmpeg")
    executor: String,
    /// Input arguments (split tokens)
    input_args: Vec<String>,
    /// Filter arguments (split tokens)
    filter_args: Vec<String>,
    /// Output file extension
    output_ext: String,
    /// Output video codec
    output_codec: String,
    /// Extra arguments (split tokens)
    extra_args: Vec<String>,
}

impl VideoPreset {
    /// Create new video preset
    #[must_use]
    pub fn new(
        executor: &str,
        input_args: &[&str],
        filter_args: &[&str],
        output_ext: &str,
        output_codec: &str,
        extra_args: &[&str],
    ) -> Self {
        Self {
            executor: executor.to_string(),
            input_args: input_args
                .iter()
                .map(std::string::ToString::to_string)
                .collect(),
            filter_args: filter_args
                .iter()
                .map(std::string::ToString::to_string)
                .collect(),
            output_ext: output_ext.to_string(),
            output_codec: output_codec.to_string(),
            extra_args: extra_args
                .iter()
                .map(std::string::ToString::to_string)
                .collect(),
        }
    }

    /// Get output file path
    fn output_path(&self, input_path: &Path) -> PathBuf {
        input_path.with_extension(&self.output_ext)
    }

    /// Build argv for processing video
    fn argv(&self, input_path: &Path, output_path: &Path) -> (String, Vec<String>) {
        let mut argv: Vec<String> = Vec::new();
        argv.extend(self.input_args.clone());
        argv.push(input_path.display().to_string());
        argv.extend(self.filter_args.clone());
        argv.push("-map_metadata".to_string());
        argv.push("0".to_string());
        argv.push("-c:v".to_string());
        argv.push(self.output_codec.clone());
        argv.extend(self.extra_args.clone());
        argv.push(output_path.display().to_string());
        (self.executor.clone(), argv)
    }
}

/// Predefined video processing preset collection
#[allow(clippy::declare_interior_mutable_const)]
pub const VIDEO_PRESETS: LazyCell<HashMap<&'static str, VideoPreset>> = LazyCell::new(|| {
    let mut map = HashMap::new();
    // 512x512 preset
    let filter_complex_512 = "[0:v]scale=512:512:force_original_aspect_ratio=increase,crop=512:512:(ow-iw)/2:(oh-ih)/2,boxblur=20[v1];[0:v]scale=512:512:force_original_aspect_ratio=decrease[v2];[v1][v2]overlay=(main_w-overlay_w)/2:(main_h-overlay_h)/2[vid]";
    map.insert(
        "AVI_512X512",
        VideoPreset::new(
            "ffmpeg",
            &["-hide_banner", "-i"],
            &["-filter_complex", filter_complex_512, "-map", "[vid]"],
            "avi",
            "mpeg4",
            &["-an", "-q:v", "8"],
        ),
    );
    map.insert(
        "WMV2_512X512",
        VideoPreset::new(
            "ffmpeg",
            &["-hide_banner", "-i"],
            &["-filter_complex", filter_complex_512, "-map", "[vid]"],
            "wmv",
            "wmv2",
            &["-an", "-q:v", "8"],
        ),
    );
    map.insert(
        "MPEG1VIDEO_512X512",
        VideoPreset::new(
            "ffmpeg",
            &["-hide_banner", "-i"],
            &["-filter_complex", filter_complex_512, "-map", "[vid]"],
            "mpg",
            "mpeg1video",
            &["-an", "-b:v", "1500k"],
        ),
    );

    // 480p preset
    let filter_complex_480 = "[0:v]scale=640:480:force_original_aspect_ratio=increase,crop=640:480:(ow-iw)/2:(oh-ih)/2,boxblur=20[v1];[0:v]scale=640:480:force_original_aspect_ratio=decrease[v2];[v1][v2]overlay=(main_w-overlay_w)/2:(main_h-overlay_h)/2[vid]";
    map.insert(
        "AVI_480P",
        VideoPreset::new(
            "ffmpeg",
            &["-hide_banner", "-i"],
            &["-filter_complex", filter_complex_480, "-map", "[vid]"],
            "avi",
            "mpeg4",
            &["-an", "-q:v", "8"],
        ),
    );
    map.insert(
        "WMV2_480P",
        VideoPreset::new(
            "ffmpeg",
            &["-hide_banner", "-i"],
            &["-filter_complex", filter_complex_480, "-map", "[vid]"],
            "wmv",
            "wmv2",
            &["-an", "-q:v", "8"],
        ),
    );
    map.insert(
        "MPEG1VIDEO_480P",
        VideoPreset::new(
            "ffmpeg",
            &["-hide_banner", "-i"],
            &["-filter_complex", filter_complex_480, "-map", "[vid]"],
            "mpg",
            "mpeg1video",
            &["-an", "-b:v", "1500k"],
        ),
    );

    map
});

/// Get media file information (using ffprobe)
///
/// # Parameters
/// - `file_path`: file path to probe
///
/// # Returns
/// Structure containing media information
async fn get_media_file_probe(file_path: &Path) -> io::Result<MediaProbe> {
    which("ffprobe").map_err(|_| io::Error::other("Executable not found: ffprobe"))?;

    let output = Command::new("ffprobe")
        .args([
            "-show_format",
            "-show_streams",
            "-print_format",
            "json",
            "-v",
            "quiet",
            &file_path.display().to_string(),
        ])
        .output()
        .await
        .map_err(|_| io::Error::other("Failed to execute ffprobe command"))?;

    if !output.status.success() {
        return Err(io::Error::other(format!(
            "ffprobe failed with status: {}\nStderr: {}",
            output.status,
            String::from_utf8_lossy(&output.stderr)
        )));
    }

    let json_str = String::from_utf8_lossy(&output.stdout);
    let probe: MediaProbe = serde_json::from_str(&json_str)
        .map_err(|_| io::Error::other("Failed to parse ffprobe JSON"))?;

    Ok(probe)
}

/// Get video information
///
/// # Parameters
/// - `file_path`: video file path
///
/// # Returns
/// Video information structure
async fn get_video_info(file_path: &Path) -> io::Result<VideoInfo> {
    let probe = get_media_file_probe(file_path).await?;

    for stream in probe.streams {
        if stream.codec_type == "video" {
            let width = stream
                .width
                .ok_or_else(|| io::Error::other("Missing width in video stream"))?;
            let height = stream
                .height
                .ok_or_else(|| io::Error::other("Missing height in video stream"))?;

            // Parse bitrate (may be string or number)
            let bit_rate = stream
                .bit_rate
                .as_ref()
                .and_then(|s| s.parse::<i32>().ok())
                .unwrap_or(0);

            return Ok(VideoInfo {
                width,
                height,
                bit_rate,
            });
        }
    }

    Err(io::Error::other("No video stream found in file"))
}

/// Get video dimensions
///
/// # Parameters
/// - `file_path`: video file path
///
/// # Returns
/// Video width and height tuple
async fn get_video_size(file_path: &Path) -> io::Result<(i32, i32)> {
    let info = get_video_info(file_path).await?;
    Ok((info.width, info.height))
}

/// Get recommended preset list (based on video aspect ratio)
///
/// # Parameters
/// - `file_path`: video file path
///
/// # Returns
/// List of recommended preset names
async fn get_preferred_presets(file_path: &Path) -> io::Result<Vec<&'static str>> {
    let (width, height) = get_video_size(file_path).await?;
    let aspect_ratio = width as f32 / height as f32;
    let target_aspect = 640.0 / 480.0; // Standard aspect ratio for 480p

    if aspect_ratio > target_aspect {
        // Use 480p preset for widescreen videos
        Ok(vec!["MPEG1VIDEO_480P", "WMV2_480P", "AVI_480P"])
    } else {
        // Use 512x512 preset for others
        Ok(vec!["MPEG1VIDEO_512X512", "WMV2_512X512", "AVI_512X512"])
    }
}

/// Process video files in directory
///
/// # Parameters
/// - `dir_path`: target directory path
/// - `input_extensions`: list of input file extensions
/// - `preset_names`: list of preset names
/// - `remove_original`: remove original file on success
/// - `remove_existing`: remove existing output files
/// - `use_preferred`: whether to use recommended presets
///
/// # Returns
/// Whether processing was successful
async fn process_videos_in_directory(
    dir_path: &Path,
    input_extensions: &[&str],
    preset_names: &[&str],
    remove_original: bool,
    remove_existing: bool,
    use_preferred: bool,
) -> io::Result<bool> {
    // Pre-check executors existence for provided presets
    {
        use std::collections::HashSet;
        let mut executors: HashSet<String> = HashSet::new();
        for name in preset_names {
            #[allow(clippy::borrow_interior_mutable_const)]
            if let Some(p) = VIDEO_PRESETS.get(*name) {
                executors.insert(p.executor.clone());
            }
        }
        for exec in executors {
            which(&exec).map_err(|_| io::Error::other(format!("Executable not found: {exec}")))?;
        }
    }
    // Collect tasks
    let mut entries = fs::read_dir(dir_path).await?;
    let mut tasks: Vec<PathBuf> = Vec::new();
    while let Some(entry) = smol::stream::StreamExt::next(&mut entries).await {
        let entry = entry?;
        let file_path = entry.path();
        if !file_path.is_file() {
            continue;
        }
        if let Some(ext) = file_path.extension().and_then(OsStr::to_str)
            && input_extensions.iter().any(|e| e.eq_ignore_ascii_case(ext))
        {
            tasks.push(file_path);
        }
    }

    let had_error = Arc::new(AtomicBool::new(false));

    // Process concurrently using disk locks（流式）
    stream::iter(tasks)
        .map(|file_path| {
            let had_error = had_error.clone();
            let preset_names = preset_names.to_vec();
            async move {
                log::info!("Processing video: {}", file_path.display());

                // Choose preset
                let mut presets_to_try = preset_names;
                if use_preferred && let Ok(preferred) = get_preferred_presets(&file_path).await {
                    presets_to_try = preferred;
                    presets_to_try.extend(presets_to_try.clone());
                }

                let mut success = false;
                for preset_name in &presets_to_try {
                    #[allow(clippy::borrow_interior_mutable_const)]
                    let Some(preset) = VIDEO_PRESETS.get(*preset_name).cloned() else {
                        continue;
                    };

                    let output_path = preset.output_path(&file_path);
                    if file_path == output_path {
                        continue;
                    }

                    if output_path.exists() {
                        if remove_existing {
                            if let Err(e) = remove_file(&output_path).await {
                                eprintln!("Failed to remove existing file: {e}");
                            }
                        } else {
                            log::info!("Output file exists, skipping: {}", output_path.display());
                            continue;
                        }
                    }

                    let (program, argv) = preset.argv(&file_path, &output_path);
                    log::info!("Executing: {} {:?}", program, argv);

                    let output = Command::new(&program).args(&argv).output().await;

                    match output {
                        Ok(output) if output.status.success() => {
                            log::info!("Successfully converted: {}", output_path.display());
                            success = true;
                            if remove_original && let Err(e) = { remove_file(&file_path).await } {
                                eprintln!("Failed to remove original file: {e}");
                            }
                            break;
                        }
                        Ok(output) => {
                            eprintln!(
                                "Conversion failed for preset {}: {}",
                                preset_name,
                                String::from_utf8_lossy(&output.stderr)
                            );
                            if output_path.exists() {
                                let _ = remove_file(&output_path).await;
                            }
                        }
                        Err(e) => {
                            eprintln!("Command execution error: {e}");
                        }
                    }
                }

                if !success {
                    had_error.store(true, Ordering::Relaxed);
                    eprintln!("All presets failed for: {}", file_path.display());
                }

                Ok::<(), io::Error>(())
            }
        })
        .buffer_unordered(64)
        .try_for_each(|_| async { Ok(()) })
        .await?;

    Ok(!had_error.load(Ordering::Relaxed))
}

/// Process all BMS folders under root directory
///
/// # Parameters
/// - `root_dir`: root directory path
/// - `input_extensions`: list of input file extensions
/// - `preset_names`: list of preset names
/// - `remove_original`: remove original file on success
/// - `remove_existing`: remove existing output files
/// - `use_preferred`: whether to use recommended presets
///
/// # Errors
///
/// Returns an error if directory operations or video processing fails
pub async fn process_bms_video_folders(
    root_dir: &Path,
    input_extensions: &[&str],
    preset_names: &[&str],
    remove_original: bool,
    remove_existing: bool,
    use_preferred: bool,
) -> io::Result<()> {
    // Validate preset names
    for name in preset_names {
        #[allow(clippy::borrow_interior_mutable_const)]
        if !VIDEO_PRESETS.contains_key(*name) {
            return Err(io::Error::other(format!("Invalid preset name: {name}")));
        }
    }

    let mut entries = fs::read_dir(root_dir).await?;
    while let Some(entry) = smol::stream::StreamExt::next(&mut entries).await {
        let entry = entry?;
        let dir_path = entry.path();
        if !dir_path.is_dir() {
            continue;
        }

        log::info!("Processing BMS folder: {}", dir_path.display());

        match process_videos_in_directory(
            &dir_path,
            input_extensions,
            preset_names,
            remove_original,
            remove_existing,
            use_preferred,
        )
        .await
        {
            Ok(true) => log::info!("Successfully processed {}", dir_path.display()),
            Ok(false) => eprintln!("Errors occurred in {}", dir_path.display()),
            Err(e) => eprintln!("Error processing {}: {}", dir_path.display(), e),
        }
    }

    Ok(())
}

// compute_parallelism_for_dir has been moved to crate::fs module

File: src/lib/utils/media/video.ts
/**
 * 视频转换模块
 * 从 Rust 代码迁移：src-tauri/src/media/video.rs
 */

import * as fs from '@tauri-apps/plugin-fs';
import { ProcessRunner } from './processRunner';
import { ConcurrencyPool } from './concurrency';
import { VIDEO_PRESETS } from './presets';
import type { VideoInfo, VideoPreset, VideoProcessParams } from './types';
import type { IProgressManager } from '../progress';
import { VIDEO_FILE_EXTS } from '../bms/scanner';

/**
 * 视频转换器类
 */
export class VideoConverter {
  /**
   * 批量处理 BMS 文件夹
   * 对应 Rust: process_bms_video_folders (video.rs:448-490)
   *
   * @param params - 视频处理参数
   * @throws 如果目录操作或视频处理失败
   */
  static async processBmsFolders(params: VideoProcessParams): Promise<void> {
    const {
      rootDir,
      inputExtensions,
      presetNames,
      removeOriginal,
      removeExisting,
      usePreferred,
      progressManager,
    } = params;

    // 启动进度管理器
    progressManager?.start();

    try {
      // 验证预设名称
      for (const name of presetNames) {
        const presetName = typeof name === 'string' ? name : String(name);
        if (!VIDEO_PRESETS[presetName]) {
          throw new Error(`Invalid preset name: ${presetName}`);
        }
      }

      // 遍历根目录下的所有子目录
      const entries = await fs.readDir(rootDir);
      const folders = [];

      for (const entry of entries) {
        // 检查是否是目录
        const entryPath = `${rootDir}/${entry.name}`;
        let isDir = false;
        try {
          const metadata = await fs.stat(entryPath);
          isDir = metadata.isDirectory ?? false;
        } catch {
          continue;
        }

        if (!isDir) continue;

        folders.push({
          name: entry.name,
          path: entryPath,
        });
      }

      // 更新总进度
      progressManager?.update(0, folders.length, `找到 ${folders.length} 个文件夹`);

      for (let i = 0; i < folders.length; i++) {
        // 检查是否应该停止（暂停或取消）
        if (progressManager?.shouldStop()) {
          if (progressManager.getProgress().cancelled) {
            console.log('任务已取消');
            return;
          }
          // 等待恢复
          await progressManager.waitForResume();
        }

        const folder = folders[i];
        const dirPath = folder.path;

        progressManager?.update(i, folders.length, `处理 ${folder.name}`);

        console.log(`Processing BMS folder: ${dirPath}`);

        try {
          const success = await this.convertInDirectory(
            dirPath,
            inputExtensions,
            presetNames.map((n) => (typeof n === 'string' ? n : String(n))),
            removeOriginal,
            removeExisting,
            usePreferred,
            progressManager
          );

          if (success) {
            console.log(`Successfully processed ${dirPath}`);
          } else {
            console.error(`Errors occurred in ${dirPath}`);
          }
        } catch (error) {
          console.error(`Error processing ${dirPath}:`, error);
          // 遇到错误时跳过，不抛出异常
        }
      }

      // 完成
      progressManager?.update(folders.length, folders.length, '视频转换完成');
    } catch (error) {
      progressManager?.reportError(error instanceof Error ? error.message : String(error));
      throw error;
    }
  }

  /**
   * 处理单个目录下的视频文件
   * 对应 Rust: process_videos_in_directory (video.rs:313-433)
   *
   * @param dirPath - 目录路径
   * @param inputExtensions - 输入文件扩展名列表
   * @param presetNames - 预设名称列表
   * @param removeOriginal - 成功时是否删除原文件
   * @param removeExisting - 是否删除已存在的输出文件
   * @param usePreferred - 是否使用推荐预设
   * @param progressManager - 进度管理器（可选）
   * @returns 是否成功
   */
  static async convertInDirectory(
    dirPath: string,
    inputExtensions: string[],
    presetNames: string[],
    removeOriginal: boolean,
    removeExisting: boolean,
    usePreferred: boolean,
    progressManager?: IProgressManager
  ): Promise<boolean> {
    // 预检查可执行文件是否存在
    await this.checkExecutables(presetNames);

    // 收集需要处理的文件
    const files = await this.collectFiles(dirPath, inputExtensions);

    const hadError = { value: false };

    // 使用并发池处理文件
    const pool = new ConcurrencyPool(64);

    for (const filePath of files) {
      // 检查是否应该停止（暂停或取消）
      if (progressManager?.shouldStop()) {
        if (progressManager.getProgress().cancelled) {
          console.log('转换已取消');
          break;
        }
        // 等待恢复
        await progressManager.waitForResume();
      }

      const fileName = filePath.split('/').pop() || filePath.split('\\').pop() || '';
      progressManager?.setMessage(`转换 ${fileName}`);

      await pool.add(async () => {
        console.log(`Processing video: ${filePath}`);

        // 选择预设
        let presetsToTry = presetNames;
        if (usePreferred) {
          try {
            const preferred = await this.getPreferredPresets(filePath);
            presetsToTry = [...preferred, ...presetNames];
          } catch (error) {
            console.error(`Failed to get preferred presets: ${error}`);
          }
        }

        let success = false;
        for (const presetName of presetsToTry) {
          const preset = VIDEO_PRESETS[presetName];
          if (!preset) continue;

          const outputPath = this.replaceExtension(filePath, preset.outputExt);

          // 跳过相同路径
          if (filePath === outputPath) continue;

          // 检查输出文件是否存在
          const outputExists = await this.fileExists(outputPath);
          if (outputExists) {
            if (removeExisting) {
              try {
                await fs.remove(outputPath);
              } catch (error) {
                console.error(`Failed to remove existing file: ${outputPath}`, error);
              }
            } else {
              console.log(`Output file exists, skipping: ${outputPath}`);
              continue;
            }
          }

          // 构建命令
          const args = this.buildCommandArgs(filePath, outputPath, preset);
          console.log(`Executing: ${preset.executor} ${args.join(' ')}`);

          const result = await ProcessRunner.exec(preset.executor, args);

          if (result.success) {
            console.log(`Successfully converted: ${outputPath}`);
            success = true;
            if (removeOriginal) {
              try {
                await fs.remove(filePath);
              } catch (error) {
                console.error(`Failed to remove original file: ${filePath}`, error);
              }
            }
            break;
          } else {
            console.error(`Conversion failed for preset ${presetName}: ${result.stderr}`);
            // 删除失败的输出文件
            if (await this.fileExists(outputPath)) {
              try {
                await fs.remove(outputPath);
              } catch (error) {
                console.error(`Failed to remove failed output: ${outputPath}`, error);
              }
            }
          }
        }

        if (!success) {
          hadError.value = true;
          console.error(`All presets failed for: ${filePath}`);
        }
      });
    }

    // 等待所有任务完成
    await pool.drain();

    return !hadError.value;
  }

  /**
   * 获取视频信息
   * 对应 Rust: get_video_info (video.rs:238-266)
   *
   * @param filePath - 视频文件路径
   * @returns 视频信息
   * @throws 如果无法获取视频信息
   */
  static async getVideoInfo(filePath: string): Promise<VideoInfo> {
    // 检查 ffprobe 是否可用
    const ffprobeExists = await ProcessRunner.checkExecutable('ffprobe');
    if (!ffprobeExists) {
      throw new Error('Executable not found: ffprobe');
    }

    const args = [
      '-show_format',
      '-show_streams',
      '-print_format',
      'json',
      '-v',
      'quiet',
      filePath,
    ];

    const result = await ProcessRunner.exec('ffprobe', args);

    if (!result.success) {
      throw new Error(`ffprobe failed: ${result.stderr}`);
    }

    try {
      const probe = JSON.parse(result.stdout);

      for (const stream of probe.streams) {
        if (stream.codec_type === 'video') {
          const width = stream.width;
          const height = stream.height;
          if (!width || !height) {
            throw new Error('Missing width or height in video stream');
          }

          // 解析比特率
          const bitRate = stream.bit_rate ? parseInt(stream.bit_rate, 10) : 0;

          return { width, height, bitRate };
        }
      }

      throw new Error('No video stream found in file');
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error('Failed to parse ffprobe JSON');
    }
  }

  /**
   * 根据视频宽高比获取推荐的预设
   * 对应 Rust: get_preferred_presets (video.rs:287-299)
   *
   * @param filePath - 视频文件路径
   * @returns 推荐的预设名称列表
   */
  static async getPreferredPresets(filePath: string): Promise<string[]> {
    const info = await this.getVideoInfo(filePath);
    const aspectRatio = info.width / info.height;
    const targetAspect = 640 / 480; // 标准 480p 宽高比

    if (aspectRatio > targetAspect) {
      // 宽屏视频使用 480p 预设
      return ['MPEG1VIDEO_480P', 'WMV2_480P', 'AVI_480P'];
    } else {
      // 其他视频使用 512x512 预设
      return ['MPEG1VIDEO_512X512', 'WMV2_512X512', 'AVI_512X512'];
    }
  }

  /**
   * 构建视频转换命令参数
   * 对应 Rust: VideoPreset::argv (video.rs:98-110)
   *
   * @param inputPath - 输入文件路径
   * @param outputPath - 输出文件路径
   * @param preset - 视频预设
   * @returns 命令参数数组
   */
  private static buildCommandArgs(
    inputPath: string,
    outputPath: string,
    preset: {
      executor: string;
      inputArgs: string[];
      filterArgs: string[];
      outputExt: string;
      outputCodec: string;
      extraArgs: string[];
    }
  ): string[] {
    return [
      ...preset.inputArgs,
      inputPath,
      ...preset.filterArgs,
      '-map_metadata',
      '0',
      '-c:v',
      preset.outputCodec,
      ...preset.extraArgs,
      outputPath,
    ];
  }

  /**
   * 收集目录中需要处理的文件
   *
   * @param dirPath - 目录路径
   * @param extensions - 文件扩展名列表
   * @returns 文件路径数组
   */
  private static async collectFiles(dirPath: string, extensions: string[]): Promise<string[]> {
    const files: string[] = [];
    const entries = await fs.readDir(dirPath);

    for (const entry of entries) {
      if (entry.name === '.' || entry.name === '..') continue;

      const filePath = `${dirPath}/${entry.name}`;

      // 检查是否是文件
      let isFile = false;
      try {
        const metadata = await fs.stat(filePath);
        isFile = metadata.isFile ?? false;
      } catch {
        continue;
      }

      if (!isFile) continue;

      // 检查文件扩展名
      const ext = this.getExtension(entry.name);
      if (ext && extensions.includes(ext.toLowerCase())) {
        files.push(filePath);
      }
    }

    return files;
  }

  /**
   * 预检查可执行文件是否存在
   *
   * @param presetNames - 预设名称列表
   * @throws 如果可执行文件不存在
   */
  private static async checkExecutables(presetNames: string[]): Promise<void> {
    const executors = new Set<string>();

    for (const name of presetNames) {
      const preset = VIDEO_PRESETS[name];
      if (preset) {
        executors.add(preset.executor);
      }
    }

    for (const executor of executors) {
      const exists = await ProcessRunner.checkExecutable(executor);
      if (!exists) {
        throw new Error(`Executable not found: ${executor}`);
      }
    }
  }

  /**
   * 替换文件扩展名
   *
   * @param filePath - 文件路径
   * @param newExt - 新扩展名（不带点）
   * @returns 新文件路径
   */
  private static replaceExtension(filePath: string, newExt: string): string {
    const lastDotIndex = filePath.lastIndexOf('.');
    if (lastDotIndex === -1) {
      return `${filePath}.${newExt}`;
    }
    return filePath.substring(0, lastDotIndex + 1) + newExt;
  }

  /**
   * 获取文件扩展名（不带点）
   *
   * @param fileName - 文件名
   * @returns 扩展名或空字符串
   */
  private static getExtension(fileName: string): string {
    const lastDotIndex = fileName.lastIndexOf('.');
    if (lastDotIndex === -1) {
      return '';
    }
    return fileName.substring(lastDotIndex + 1);
  }

  /**
   * 检查文件是否存在
   *
   * @param filePath - 文件路径
   * @returns 是否存在
   */
  private static async fileExists(filePath: string): Promise<boolean> {
    try {
      await fs.stat(filePath);
      return true;
    } catch {
      return false;
    }
  }
}

/**
 * 视频文件转换
 * 对应 Python: transfer_video (bms_folder_media.py:35-52)
 *
 * @command
 * @category media
 * @dangerous true
 * @name 视频文件转换
 * @description 转换BMS根目录下的视频文件
 * @frontend true
 *
 * @param {string} rootDir - 根目录路径
 * @param {VideoPreset[]} presetNames - 目标格式预设名称
 * @param {boolean} removeOriginFile - 成功时删除原文件
 * @param {boolean} removeExistingTargetFile - 删除已存在的目标文件
 * @param {boolean} usePrefered - 使用推荐预设
 * @param {IProgressManager} progressManager - 进度管理器（可选）
 *
 * @returns {Promise<void>}
 */
export async function transferVideo(
  rootDir: string,
  presetNames: VideoPreset[],
  removeOriginFile: boolean,
  removeExistingTargetFile: boolean,
  usePrefered: boolean,
  progressManager?: IProgressManager
): Promise<void> {
  await VideoConverter.processBmsFolders({
    rootDir,
    inputExtensions: [...VIDEO_FILE_EXTS],
    presetNames,
    removeOriginal: removeOriginFile,
    removeExisting: removeExistingTargetFile,
    usePreferred: usePrefered,
    progressManager,
  });
}
